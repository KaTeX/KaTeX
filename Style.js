function Style(id, size, multiplier, cramped) {
    this.id = id;
    this.size = size;
    this.cramped = cramped;
    this.sizeMultiplier = multiplier;
}

Style.prototype.sup = function() {
    return styles[sup[this.id]];
};

Style.prototype.sub = function() {
    return styles[sub[this.id]];
};

Style.prototype.fracNum = function() {
    return styles[fracNum[this.id]];
};

Style.prototype.fracDen = function() {
    return styles[fracDen[this.id]];
};

Style.prototype.cramp = function() {
    return styles[cramp[this.id]];
};

// HTML class name, like "displaystyle cramped"
Style.prototype.cls = function() {
    return sizeNames[this.size] + (this.cramped ? " cramped" : " uncramped");
};

// HTML Reset class name, like "reset-textstyle"
Style.prototype.reset = function() {
    return resetNames[this.size];
};

var D = 0;
var Dc = 1;
var T = 2;
var Tc = 3;
var S = 4;
var Sc = 5;
var SS = 6;
var SSc = 7;

var sizeNames = [
    "displaystyle textstyle",
    "textstyle",
    "scriptstyle",
    "scriptscriptstyle"
];

var resetNames = [
    "reset-textstyle",
    "reset-textstyle",
    "reset-scriptstyle",
    "reset-scriptscriptstyle"
];

var styles = [
    new Style(D, 0, 1.0, false),
    new Style(Dc, 0, 1.0, true),
    new Style(T, 1, 1.0, false),
    new Style(Tc, 1, 1.0, true),
    new Style(S, 2, 0.66667, false),
    new Style(Sc, 2, 0.66667, true),
    new Style(SS, 3, 0.5, false),
    new Style(SSc, 3, 0.5, true)
];

var sup = [S, Sc, S, Sc, SS, SSc, SS, SSc];
var sub = [Sc, Sc, Sc, Sc, SSc, SSc, SSc, SSc];
var fracNum = [T, Tc, S, Sc, SS, SSc, SS, SSc];
var fracDen = [Tc, Tc, Sc, Sc, SSc, SSc, SSc, SSc];
var cramp = [Dc, Dc, Tc, Tc, Sc, Sc, SSc, SSc];

module.exports = {
    DISPLAY: styles[D],
    TEXT: styles[T]
};
