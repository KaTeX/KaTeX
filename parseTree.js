var Parser = require("./Parser");
var parser = new Parser();

var parseTree = function(toParse) {
    return parser.parse(toParse);
};

module.exports = parseTree;
