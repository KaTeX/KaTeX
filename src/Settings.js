// @flow
/* eslint no-console:0 */
/**
 * This is a module for storing settings passed into KaTeX. It correctly handles
 * default settings.
 */

import utils from "./utils";
import ParseError from "./ParseError";
import {Token} from "./Token";

import type {AnyParseNode} from "./parseNode";
import type {MacroMap} from "./macros";

export type StrictFunction =
    (errorCode: string, errorMsg: string, token?: Token | AnyParseNode) =>
    ?(boolean | string);

export type SettingsOptions = {
    displayMode?: boolean;
    leqno?: boolean;
    fleqn?: boolean;
    throwOnError?: boolean;
    errorColor?: string;
    macros?: MacroMap;
    colorIsTextColor?: boolean;
    strict?: boolean | "ignore" | "warn" | "error" | StrictFunction;
    maxSize?: number;
    maxExpand?: number;
    allowedProtocols?: string[];
};

/**
 * The main Settings object
 *
 * The current options stored are:
 *  - displayMode: Whether the expression should be typeset as inline math
 *                 (false, the default), meaning that the math starts in
 *                 \textstyle and is placed in an inline-block); or as display
 *                 math (true), meaning that the math starts in \displaystyle
 *                 and is placed in a block with vertical margin.
 */
class Settings {
    displayMode: boolean;
    leqno: boolean;
    fleqn: boolean;
    throwOnError: boolean;
    errorColor: string;
    macros: MacroMap;
    colorIsTextColor: boolean;
    strict: boolean | "ignore" | "warn" | "error" | StrictFunction;
    maxSize: number;
    maxExpand: number;
    allowedProtocols: string[];

    constructor(options: SettingsOptions) {
        // allow null options
        options = options || {};
        this.displayMode = utils.deflt(options.displayMode, false);
        this.leqno = utils.deflt(options.leqno, false);
        this.fleqn = utils.deflt(options.fleqn, false);
        this.throwOnError = utils.deflt(options.throwOnError, true);
        this.errorColor = utils.deflt(options.errorColor, "#cc0000");
        this.macros = options.macros || {};
        this.colorIsTextColor = utils.deflt(options.colorIsTextColor, false);
        this.strict = utils.deflt(options.strict, "warn");
        this.maxSize = Math.max(0, utils.deflt(options.maxSize, Infinity));
        this.maxExpand = Math.max(0, utils.deflt(options.maxExpand, 1000));
        this.allowedProtocols = utils.deflt(options.allowedProtocols,
            ["http", "https", "mailto", "_relative"]);
    }

    /**
     * Report nonstrict (non-LaTeX-compatible) input.
     * Can safely not be called if `this.strict` is false in JavaScript.
     */
    reportNonstrict(errorCode: string, errorMsg: string,
                    token?: Token | AnyParseNode) {
        let strict = this.strict;
        if (typeof strict === "function") {
            // Allow return value of strict function to be boolean or string
            // (or null/undefined, meaning no further processing).
            strict = strict(errorCode, errorMsg, token);
        }
        if (!strict || strict === "ignore") {
            return;
        } else if (strict === true || strict === "error") {
            throw new ParseError(
                "LaTeX-incompatible input and strict mode is set to 'error': " +
                `${errorMsg} [${errorCode}]`, token);
        } else if (strict === "warn") {
            typeof console !== "undefined" && console.warn(
                "LaTeX-incompatible input and strict mode is set to 'warn': " +
                `${errorMsg} [${errorCode}]`);
        } else {  // won't happen in type-safe code
            typeof console !== "undefined" && console.warn(
                "LaTeX-incompatible input and strict mode is set to " +
                `unrecognized '${strict}': ${errorMsg} [${errorCode}]`);
        }
    }

    /**
     * Check whether to apply strict (LaTeX-adhering) behavior for unusual
     * input (like `\\`).  Unlike `nonstrict`, will not throw an error;
     * instead, "error" translates to a return value of `true`, while "ignore"
     * translates to a return value of `false`.  May still print a warning:
     * "warn" prints a warning and returns `false`.
     * This is for the second category of `errorCode`s listed in the README.
     */
    useStrictBehavior(errorCode: string, errorMsg: string,
                      token?: Token | AnyParseNode) {
        let strict = this.strict;
        if (typeof strict === "function") {
            // Allow return value of strict function to be boolean or string
            // (or null/undefined, meaning no further processing).
            // But catch any exceptions thrown by function, treating them
            // like "error".
            try {
                strict = strict(errorCode, errorMsg, token);
            } catch (error) {
                strict = "error";
            }
        }
        if (!strict || strict === "ignore") {
            return false;
        } else if (strict === true || strict === "error") {
            return true;
        } else if (strict === "warn") {
            typeof console !== "undefined" && console.warn(
                "LaTeX-incompatible input and strict mode is set to 'warn': " +
                `${errorMsg} [${errorCode}]`);
            return false;
        } else {  // won't happen in type-safe code
            typeof console !== "undefined" && console.warn(
                "LaTeX-incompatible input and strict mode is set to " +
                `unrecognized '${strict}': ${errorMsg} [${errorCode}]`);
            return false;
        }
    }
}

export default Settings;
