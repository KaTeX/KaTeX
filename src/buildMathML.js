// @flow
/**
 * This file converts a parse tree into a cooresponding MathML tree. The main
 * entry point is the `buildMathML` function, which takes a parse tree from the
 * parser.
 */

import buildCommon from "./buildCommon";
import {getCharacterMetrics} from "./fontMetrics";
import mathMLTree from "./mathMLTree";
import ParseError from "./ParseError";
import symbols, {ligatures} from "./symbols";
import utils from "./utils";
import {_mathmlGroupBuilders as groupBuilders} from "./defineFunction";
import {MathNode, TextNode} from "./mathMLTree";

import type Options from "./Options";
import type {AnyParseNode, SymbolParseNode} from "./parseNode";
import type {DomSpan} from "./domTree";
import type {MathDomNode} from "./mathMLTree";
import type {FontVariant, Mode} from "./types";

/**
 * Takes a symbol and converts it into a MathML text node after performing
 * optional replacement from symbols.js.
 */
export const makeText = function(
    text: string,
    mode: Mode,
    options?: Options,
): TextNode {
    if (symbols[mode][text] && symbols[mode][text].replace &&
        text.charCodeAt(0) !== 0xD835 &&
        !(ligatures.hasOwnProperty(text) && options &&
          ((options.fontFamily && options.fontFamily.substr(4, 2) === "tt") ||
           (options.font && options.font.substr(4, 2) === "tt")))) {
        text = symbols[mode][text].replace;
    }

    return new mathMLTree.TextNode(text);
};

/**
 * Wrap the given array of nodes in an <mrow> node if needed, i.e.,
 * unless the array has length 1.  Always returns a single node.
 */
export const makeRow = function(body: MathDomNode[]): MathDomNode {
    if (body.length === 1) {
        return body[0];
    } else {
        return new mathMLTree.MathNode("mrow", body);
    }
};

/**
 * Returns the math variant as a string or null if none is required.
 */
export const getVariant = function(
    group: SymbolParseNode,
    options: Options,
): ?FontVariant {
    // Handle \text... font specifiers as best we can.
    // MathML has a limited list of allowable mathvariant specifiers; see
    // https://www.w3.org/TR/MathML3/chapter3.html#presm.commatt
    if (options.fontFamily === "texttt") {
        return "monospace";
    } else if (options.fontFamily === "textsf") {
        if (options.fontShape === "textit" &&
            options.fontWeight === "textbf") {
            return "sans-serif-bold-italic";
        } else if (options.fontShape === "textit") {
            return "sans-serif-italic";
        } else if (options.fontWeight === "textbf") {
            return "bold-sans-serif";
        } else {
            return "sans-serif";
        }
    } else if (options.fontShape === "textit" &&
               options.fontWeight === "textbf") {
        return "bold-italic";
    } else if (options.fontShape === "textit") {
        return "italic";
    } else if (options.fontWeight === "textbf") {
        return "bold";
    }

    const font = options.font;
    if (!font || font === "mathnormal") {
        return null;
    }

    const mode = group.mode;
    if (font === "mathit") {
        return "italic";
    } else if (font === "boldsymbol") {
        return "bold-italic";
    }

    let text = group.text;
    if (utils.contains(["\\imath", "\\jmath"], text)) {
        return null;
    }

    if (symbols[mode][text] && symbols[mode][text].replace) {
        text = symbols[mode][text].replace;
    }

    const fontName = buildCommon.fontMap[font].fontName;
    if (getCharacterMetrics(text, fontName, mode)) {
        return buildCommon.fontMap[font].variant;
    }

    return null;
};

/**
 * Takes a list of nodes, builds them, and returns a list of the generated
 * MathML nodes.  Also combine consecutive <mtext> outputs into a single
 * <mtext> tag.
 */
export const buildExpression = function(
    expression: AnyParseNode[],
    options: Options,
): MathDomNode[] {
    const groups = [];
    let lastGroup;
    for (let i = 0; i < expression.length; i++) {
        const group = buildGroup(expression[i], options);
        if (group instanceof MathNode && lastGroup instanceof MathNode) {
            // Concatenate adjacent <mtext>s
            if (group.type === 'mtext' && lastGroup.type === 'mtext'
                && group.getAttribute('mathvariant') ===
                   lastGroup.getAttribute('mathvariant')) {
                lastGroup.children.push(...group.children);
                continue;
            // Concatenate adjacent <mn>s
            } else if (group.type === 'mn' && lastGroup.type === 'mn') {
                lastGroup.children.push(...group.children);
                continue;
            // Concatenate <mn>...</mn> followed by <mi>.</mi>
            } else if (group.type === 'mi' && group.children.length === 1 &&
                       lastGroup.type === 'mn') {
                const child = group.children[0];
                if (child instanceof TextNode && child.text === '.') {
                    lastGroup.children.push(...group.children);
                    continue;
                }
            }
        }
        groups.push(group);
        lastGroup = group;
    }

    // TODO(kevinb): combine \\not with mrels and mords

    return groups;
};

/**
 * Equivalent to buildExpression, but wraps the elements in an <mrow>
 * if there's more than one.  Returns a single node instead of an array.
 */
export const buildExpressionRow = function(
    expression: AnyParseNode[],
    options: Options,
): MathDomNode {
    return makeRow(buildExpression(expression, options));
};

/**
 * Takes a group from the parser and calls the appropriate groupBuilders function
 * on it to produce a MathML node.
 */
export const buildGroup = function(
    group: ?AnyParseNode,
    options: Options,
): MathDomNode {
    if (!group) {
        return new mathMLTree.MathNode("mrow");
    }

    if (groupBuilders[group.type]) {
        // Call the groupBuilders function
        // $FlowFixMe
        const result: MathDomNode = groupBuilders[group.type](group, options);
        return result;
    } else {
        throw new ParseError(
            "Got group of unknown type: '" + group.type + "'");
    }
};

/**
 * Takes a full parse tree and settings and builds a MathML representation of
 * it. In particular, we put the elements from building the parse tree into a
 * <semantics> tag so we can also include that TeX source as an annotation.
 *
 * Note that we actually return a domTree element with a `<math>` inside it so
 * we can do appropriate styling.
 */
export default function buildMathML(
    tree: AnyParseNode[],
    texExpression: string,
    options: Options,
): DomSpan {
    const expression = buildExpression(tree, options);

    // Wrap up the expression in an mrow so it is presented in the semantics
    // tag correctly, unless it's a single <mrow> or <mtable>.
    let wrapper;
    if (expression.length === 1 && expression[0] instanceof MathNode &&
        utils.contains(["mrow", "mtable"], expression[0].type)) {
        wrapper = expression[0];
    } else {
        wrapper = new mathMLTree.MathNode("mrow", expression);
    }

    // Build a TeX annotation of the source
    const annotation = new mathMLTree.MathNode(
        "annotation", [new mathMLTree.TextNode(texExpression)]);

    annotation.setAttribute("encoding", "application/x-tex");

    const semantics = new mathMLTree.MathNode(
        "semantics", [wrapper, annotation]);

    const math = new mathMLTree.MathNode("math", [semantics]);

    // You can't style <math> nodes, so we wrap the node in a span.
    // NOTE: The span class is not typed to have <math> nodes as children, and
    // we don't want to make the children type more generic since the children
    // of span are expected to have more fields in `buildHtml` contexts.
    // $FlowFixMe
    return buildCommon.makeSpan(["katex-mathml"], [math]);
}
