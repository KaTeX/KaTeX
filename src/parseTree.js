// @flow
/**
 * Provides a single function for parsing an expression using a Parser
 * TODO(emily): Remove this
 */

import Parser from "./Parser";
import ParseError from "./ParseError";
import ParseNode from "./ParseNode";

import type Settings from "./Settings";
import type {AnyParseNode} from "./ParseNode";

/**
 * Parses an expression using a Parser, then returns the parsed result.
 */
const parseTree = function(toParse: string, settings: Settings): AnyParseNode[] {
    if (!(typeof toParse === 'string' || toParse instanceof String)) {
        throw new TypeError('KaTeX can only parse string typed expression');
    }
    const parser = new Parser(toParse, settings);
    // Blank out any \df@tag to avoid spurious "Duplicate \tag" errors
    delete parser.gullet.macros.current["\\df@tag"];
    let tree = parser.parse();

    // If the input used \tag, it will set the \df@tag macro to the tag.
    // In this case, we separately parse the tag and wrap the tree.
    if (parser.gullet.macros.get("\\df@tag")) {
        if (!settings.displayMode) {
            throw new ParseError("\\tag works only in display equations");
        }
        parser.gullet.feed("\\df@tag");
        tree = [new ParseNode("tag", {
            type: "tag",
            body: tree,
            tag: parser.parse(),
        }, "text")];
    }

    return tree;
};

export default parseTree;
