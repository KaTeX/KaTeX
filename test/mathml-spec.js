/* global expect: false */
/* global it: false */
/* global describe: false */

import buildMathML from "../src/buildMathML";
import parseTree from "../src/parseTree";
import Options from "../src/Options";
import Settings from "../src/Settings";
import Style from "../src/Style";

const defaultSettings = new Settings({});

const getMathML = function(expr, settings) {
    const usedSettings = settings ? settings : defaultSettings;

    let startStyle = Style.TEXT;
    if (usedSettings.displayMode) {
        startStyle = Style.DISPLAY;
    }

    // Setup the default options
    const options = new Options({
        style: startStyle,
        maxSize: Infinity,
    });

    const built = buildMathML(parseTree(expr, usedSettings), expr, options);

    // Strip off the surrounding <span>
    return built.children[0].toMarkup();
};

describe("A MathML builder", function() {
    it('should generate the right types of nodes', () => {
        expect(getMathML("\\sin{x}+1\\;\\text{a}")).toMatchSnapshot();
    });

    it('should make prime operators into <mo> nodes', () => {
        expect(getMathML("f'")).toMatchSnapshot();
    });

    it('should generate <mphantom> nodes for \\phantom', () => {
        expect(getMathML("\\phantom{x}")).toMatchSnapshot();
    });

    it('should use <munderover> for large operators', () => {
        expect(getMathML("\\displaystyle\\sum_a^b")).toMatchSnapshot();
    });

    it('should use <msupsub> for regular operators', () => {
        expect(getMathML("\\textstyle\\sum_a^b")).toMatchSnapshot();
    });

    it('should use <mpadded> for raisebox', () => {
        expect(getMathML("\\raisebox{0.25em}{b}")).toMatchSnapshot();
    });

    it('should use <menclose> for colorbox', () => {
        expect(getMathML("\\colorbox{red}{b}")).toMatchSnapshot();
    });

    it('should render mathchoice as if there was nothing', () => {
        const cmd = "\\sum_{k = 0}^{\\infty} x^k";
        expect(getMathML(`\\displaystyle\\mathchoice{${cmd}}{T}{S}{SS}`))
            .toMatchSnapshot();
        expect(getMathML(`\\mathchoice{D}{${cmd}}{S}{SS}`))
            .toMatchSnapshot();
        expect(getMathML(`x_{\\mathchoice{D}{T}{${cmd}}{SS}}`))
            .toMatchSnapshot();
        expect(getMathML(`x_{y_{\\mathchoice{D}{T}{S}{${cmd}}}}`))
            .toMatchSnapshot();
    });
});
