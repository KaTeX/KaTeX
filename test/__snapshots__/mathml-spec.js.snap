// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A MathML builder \\html@mathml makes clean symbols 1`] = `

<math>
  <semantics>
    <mrow>
      <mtext>
        ©
      </mtext>
      <mi mathvariant="normal">
        ≠
      </mi>
      <mi mathvariant="normal">
        ∉
      </mi>
      <mi mathvariant="normal">
        ≘
      </mi>
      <mtext>
        KaTeX
      </mtext>
    </mrow>
    <annotation encoding="application/x-tex">
      \\copyright\\neq\\notin≘\\KaTeX
    </annotation>
  </semantics>
</math>

`;

exports[`A MathML builder \\text fonts become mathvariant 1`] = `

<math>
  <semantics>
    <mrow>
      <mtext>
        roman
      </mtext>
      <mrow>
        <mtext mathvariant="italic">
          italic
        </mtext>
        <mrow>
          <mtext mathvariant="bold-italic">
            bold
          </mtext>
          <mtext>
             
          </mtext>
          <mtext mathvariant="bold-italic">
            italic
          </mtext>
        </mrow>
      </mrow>
      <mtext mathvariant="bold">
        bold
      </mtext>
      <mrow>
        <mtext mathvariant="sans-serif">
          ss
        </mtext>
        <mrow>
          <mtext mathvariant="sans-serif-italic">
            italic
          </mtext>
          <mrow>
            <mtext mathvariant="sans-serif-bold-italic">
              bold
            </mtext>
            <mtext>
               
            </mtext>
            <mtext mathvariant="sans-serif-bold-italic">
              italic
            </mtext>
          </mrow>
        </mrow>
        <mtext mathvariant="bold-sans-serif">
          bold
        </mtext>
      </mrow>
      <mrow>
        <mtext mathvariant="monospace">
          tt
        </mtext>
        <mrow>
          <mtext mathvariant="monospace">
            italic
          </mtext>
          <mrow>
            <mtext mathvariant="monospace">
              bold
            </mtext>
            <mtext>
               
            </mtext>
            <mtext mathvariant="monospace">
              italic
            </mtext>
          </mrow>
        </mrow>
        <mtext mathvariant="monospace">
          bold
        </mtext>
      </mrow>
    </mrow>
    <annotation encoding="application/x-tex">
      \\text{roman\\textit{italic\\textbf{bold italic}}\\textbf{bold}\\textsf{ss\\textit{italic\\textbf{bold italic}}\\textbf{bold}}\\texttt{tt\\textit{italic\\textbf{bold italic}}\\textbf{bold}}}
    </annotation>
  </semantics>
</math>

`;

exports[`A MathML builder accents turn into <mover accent="true"> in MathML 1`] = `

<math>
  <semantics>
    <mrow>
      <mover accent="true">
        <mi>
          u
        </mi>
        <mo>
          ¨
        </mo>
      </mover>
      <mi>
        b
      </mi>
      <mi>
        e
      </mi>
      <mi>
        r
      </mi>
      <mi>
        f
      </mi>
      <mi>
        i
      </mi>
      <mi>
        a
      </mi>
      <mi>
        n
      </mi>
      <mi>
        c
      </mi>
      <mover accent="true">
        <mi>
          e
        </mi>
        <mo>
          ˊ
        </mo>
      </mover>
      <mi>
        e
      </mi>
    </mrow>
    <annotation encoding="application/x-tex">
      über fiancée
    </annotation>
  </semantics>
</math>

`;

exports[`A MathML builder ligatures render properly 1`] = `

<math>
  <semantics>
    <mrow>
      <mtext>
        “‘Hi—-”’
      </mtext>
      <mo>
        −
      </mo>
      <mo>
        −
      </mo>
      <mtext mathvariant="monospace">
        \`\`‘Hi----&#x27;&#x27;’
      </mtext>
      <mrow>
        <mtext>
          \`\`
        </mtext>
        <mtext mathvariant="monospace">
          ‘Hi
        </mtext>
        <mtext>
          ---
        </mtext>
        <mtext mathvariant="monospace">
          -
        </mtext>
        <mtext>
          &#x27;&#x27;
        </mtext>
        <mtext mathvariant="monospace">
          ’
        </mtext>
      </mrow>
    </mrow>
    <annotation encoding="application/x-tex">
      \\text{\`\`\`Hi----&#x27;&#x27;&#x27;}--\\texttt{\`\`\`Hi----&#x27;&#x27;&#x27;}\\text{\\tt \`\`\`Hi----&#x27;&#x27;&#x27;}
    </annotation>
  </semantics>
</math>

`;

exports[`A MathML builder normal spaces render normally 1`] = `

<math>
  <semantics>
    <mrow>
      <mspace width="1em">
      </mspace>
      <mspace width="0.431em">
      </mspace>
    </mrow>
    <annotation encoding="application/x-tex">
      \\kern1em\\kern1ex
    </annotation>
  </semantics>
</math>

`;

exports[`A MathML builder should concatenate digits into single <mn> 1`] = `

<math>
  <semantics>
    <mrow>
      <mi>
        sin
      </mi>
      <mo>
        ⁡
      </mo>
      <mi>
        α
      </mi>
      <mo>
        =
      </mo>
      <mn>
        0.34
      </mn>
    </mrow>
    <annotation encoding="application/x-tex">
      \\sin{\\alpha}=0.34
    </annotation>
  </semantics>
</math>

`;

exports[`A MathML builder should generate <mphantom> nodes for \\phantom 1`] = `

<math>
  <semantics>
    <mrow>
      <mphantom>
        <mi>
          x
        </mi>
      </mphantom>
    </mrow>
    <annotation encoding="application/x-tex">
      \\phantom{x}
    </annotation>
  </semantics>
</math>

`;

exports[`A MathML builder should generate the right types of nodes 1`] = `

<math>
  <semantics>
    <mrow>
      <mi>
        sin
      </mi>
      <mo>
        ⁡
      </mo>
      <mi>
        x
      </mi>
      <mo>
        +
      </mo>
      <mn>
        1
      </mn>
      <mtext>
          
      </mtext>
      <mtext>
        a
      </mtext>
    </mrow>
    <annotation encoding="application/x-tex">
      \\sin{x}+1\\;\\text{a}
    </annotation>
  </semantics>
</math>

`;

exports[`A MathML builder should make prime operators into <mo> nodes 1`] = `

<math>
  <semantics>
    <mrow>
      <msup>
        <mi>
          f
        </mi>
        <mo mathvariant="normal">
          ′
        </mo>
      </msup>
    </mrow>
    <annotation encoding="application/x-tex">
      f&#x27;
    </annotation>
  </semantics>
</math>

`;

exports[`A MathML builder should output \\limsup_{x \\rightarrow \\infty} correctly in \\textstyle 1`] = `

<math>
  <semantics>
    <mrow>
      <munder>
        <mo>
          <mi mathvariant="normal">
            lim sup
          </mi>
          <mo>
            ⁡
          </mo>
        </mo>
        <mrow>
          <mi>
            x
          </mi>
          <mo>
            →
          </mo>
          <mi mathvariant="normal">
            ∞
          </mi>
        </mrow>
      </munder>
    </mrow>
    <annotation encoding="application/x-tex">
      \\limsup_{x \\rightarrow \\infty}
    </annotation>
  </semantics>
</math>

`;

exports[`A MathML builder should output \\limsup_{x \\rightarrow \\infty} in displaymode correctly 1`] = `

<math>
  <semantics>
    <mrow>
      <munder>
        <mo>
          <mi mathvariant="normal">
            lim sup
          </mi>
          <mo>
            ⁡
          </mo>
        </mo>
        <mrow>
          <mi>
            x
          </mi>
          <mo>
            →
          </mo>
          <mi mathvariant="normal">
            ∞
          </mi>
        </mrow>
      </munder>
    </mrow>
    <annotation encoding="application/x-tex">
      \\limsup_{x \\rightarrow \\infty}
    </annotation>
  </semantics>
</math>

`;

exports[`A MathML builder should render boldsymbol with the correct mathvariants 1`] = `

<math>
  <semantics>
    <mrow>
      <mrow>
        <mi mathvariant="bold-italic">
          A
        </mi>
        <mi mathvariant="bold-italic">
          x
        </mi>
        <mn mathvariant="bold-italic">
          2
        </mn>
        <mi mathvariant="bold-italic">
          k
        </mi>
        <mi mathvariant="bold-italic">
          ω
        </mi>
        <mi mathvariant="bold-italic">
          Ω
        </mi>
        <mi mathvariant="bold-italic">
          ı
        </mi>
        <mo mathvariant="bold-italic">
          +
        </mo>
      </mrow>
    </mrow>
    <annotation encoding="application/x-tex">
      \\boldsymbol{Ax2k\\omega\\Omega\\imath+}
    </annotation>
  </semantics>
</math>

`;

exports[`A MathML builder should render mathchoice as if there was nothing 1`] = `

<math>
  <semantics>
    <mrow>
      <mstyle scriptlevel="0"
              displaystyle="true"
      >
        <mrow>
          <munderover>
            <mo>
              ∑
            </mo>
            <mrow>
              <mi>
                k
              </mi>
              <mo>
                =
              </mo>
              <mn>
                0
              </mn>
            </mrow>
            <mi mathvariant="normal">
              ∞
            </mi>
          </munderover>
          <msup>
            <mi>
              x
            </mi>
            <mi>
              k
            </mi>
          </msup>
        </mrow>
      </mstyle>
    </mrow>
    <annotation encoding="application/x-tex">
      \\displaystyle\\mathchoice{\\sum_{k = 0}^{\\infty} x^k}{T}{S}{SS}
    </annotation>
  </semantics>
</math>

`;

exports[`A MathML builder should render mathchoice as if there was nothing 2`] = `

<math>
  <semantics>
    <mrow>
      <msubsup>
        <mo>
          ∑
        </mo>
        <mrow>
          <mi>
            k
          </mi>
          <mo>
            =
          </mo>
          <mn>
            0
          </mn>
        </mrow>
        <mi mathvariant="normal">
          ∞
        </mi>
      </msubsup>
      <msup>
        <mi>
          x
        </mi>
        <mi>
          k
        </mi>
      </msup>
    </mrow>
    <annotation encoding="application/x-tex">
      \\mathchoice{D}{\\sum_{k = 0}^{\\infty} x^k}{S}{SS}
    </annotation>
  </semantics>
</math>

`;

exports[`A MathML builder should render mathchoice as if there was nothing 3`] = `

<math>
  <semantics>
    <mrow>
      <msub>
        <mi>
          x
        </mi>
        <mi>
          T
        </mi>
      </msub>
    </mrow>
    <annotation encoding="application/x-tex">
      x_{\\mathchoice{D}{T}{\\sum_{k = 0}^{\\infty} x^k}{SS}}
    </annotation>
  </semantics>
</math>

`;

exports[`A MathML builder should render mathchoice as if there was nothing 4`] = `

<math>
  <semantics>
    <mrow>
      <msub>
        <mi>
          x
        </mi>
        <msub>
          <mi>
            y
          </mi>
          <mi>
            T
          </mi>
        </msub>
      </msub>
    </mrow>
    <annotation encoding="application/x-tex">
      x_{y_{\\mathchoice{D}{T}{S}{\\sum_{k = 0}^{\\infty} x^k}}}
    </annotation>
  </semantics>
</math>

`;

exports[`A MathML builder should set href attribute for href appropriately 1`] = `

<math>
  <semantics>
    <mrow>
      <mi href="http://example.org">
        α
      </mi>
    </mrow>
    <annotation encoding="application/x-tex">
      \\href{http://example.org}{\\alpha}
    </annotation>
  </semantics>
</math>

`;

exports[`A MathML builder should use <menclose> for colorbox 1`] = `

<math>
  <semantics>
    <mrow>
      <menclose mathbackground="red">
        <mtext>
          b
        </mtext>
      </menclose>
    </mrow>
    <annotation encoding="application/x-tex">
      \\colorbox{red}{b}
    </annotation>
  </semantics>
</math>

`;

exports[`A MathML builder should use <mpadded> for raisebox 1`] = `

<math>
  <semantics>
    <mrow>
      <mpadded voffset="0.25em">
        <mtext>
          b
        </mtext>
      </mpadded>
    </mrow>
    <annotation encoding="application/x-tex">
      \\raisebox{0.25em}{b}
    </annotation>
  </semantics>
</math>

`;

exports[`A MathML builder should use <msupsub> for regular operators 1`] = `

<math>
  <semantics>
    <mrow>
      <mstyle scriptlevel="0"
              displaystyle="false"
      >
        <msubsup>
          <mo>
            ∑
          </mo>
          <mi>
            a
          </mi>
          <mi>
            b
          </mi>
        </msubsup>
      </mstyle>
    </mrow>
    <annotation encoding="application/x-tex">
      \\textstyle\\sum_a^b
    </annotation>
  </semantics>
</math>

`;

exports[`A MathML builder should use <munderover> for large operators 1`] = `

<math>
  <semantics>
    <mrow>
      <mstyle scriptlevel="0"
              displaystyle="true"
      >
        <munderover>
          <mo>
            ∑
          </mo>
          <mi>
            a
          </mi>
          <mi>
            b
          </mi>
        </munderover>
      </mstyle>
    </mrow>
    <annotation encoding="application/x-tex">
      \\displaystyle\\sum_a^b
    </annotation>
  </semantics>
</math>

`;

exports[`A MathML builder special spaces render specially 1`] = `

<math>
  <semantics>
    <mrow>
      <mtext>
         
      </mtext>
      <mtext>
         
      </mtext>
      <mtext>
         
      </mtext>
      <mtext>
         
      </mtext>
      <mtext>
         
      </mtext>
      <mtext>
          
      </mtext>
      <mtext>
          
      </mtext>
      <mtext>
         ⁣
      </mtext>
      <mtext>
         ⁣
      </mtext>
      <mtext>
         ⁣
      </mtext>
      <mtext>
         ⁣
      </mtext>
      <mtext>
         
      </mtext>
      <mtext>
         
      </mtext>
      <mtext>
         
      </mtext>
      <mtext>
          
      </mtext>
      <mtext>
         ⁣
      </mtext>
      <mtext>
         ⁣
      </mtext>
      <mtext>
         ⁣
      </mtext>
      <mtext>
         ⁣
      </mtext>
    </mrow>
    <annotation encoding="application/x-tex">
      \\,\\thinspace\\:\\&gt;\\medspace\\;\\thickspace\\!\\negthinspace\\negmedspace\\negthickspace\\mkern1mu\\mkern3mu\\mkern4mu\\mkern5mu\\mkern-1mu\\mkern-3mu\\mkern-4mu\\mkern-5mu
    </annotation>
  </semantics>
</math>

`;

exports[`A MathML builder tags use <mlabeledtr> 1`] = `

<math>
  <semantics>
    <mtable width="100%">
      <mtr>
        <mtd width="50%">
        </mtd>
        <mtd>
          <mrow>
            <mi>
              x
            </mi>
            <mo>
              +
            </mo>
            <msup>
              <mi>
                y
              </mi>
              <mn>
                2
              </mn>
            </msup>
          </mrow>
        </mtd>
        <mtd width="50%">
        </mtd>
        <mtd>
          <mtext>
            (hi)
          </mtext>
        </mtd>
      </mtr>
    </mtable>
    <annotation encoding="application/x-tex">
      \\tag{hi} x+y^2
    </annotation>
  </semantics>
</math>

`;
